### PR TITLE
Nvme dynamic flbas detection

### PR DESCRIPTION
**Pull request by maramsmurthy**
This commit implements dynamic FLBAS (Formatted LBA Size) detection
for NVMe namespace creation, addressing failures on devices with
non-standard FLBAS configurations.

Problem:
--------
The hardcoded --flbas=0 parameter in namespace creation was causing
failures on NVMe devices that don't support FLBAS index 0 as their
default or only supported format. This affected devices with:
- Non-standard LBA format configurations
- 512B-only or 4KB-only block size support
- Vendor-specific FLBAS arrangements

Solution:
---------
Implemented three-mode FLBAS selection strategy in create_one_ns():

1. Default Mode (flbas=0): Backward compatible
   - Attempts creation with FLBAS=0 first
   - Auto-detects optimal FLBAS on failure and retries
   - Maintains compatibility with existing test workflows

2. Auto-Detect Mode (flbas=-1): Skip FLBAS=0 attempt
   - Immediately queries and uses optimal FLBAS
   - Efficient for known non-standard devices
   - Reduces unnecessary failed attempts

3. Explicit Mode (flbas=1-15): User-specified FLBAS
   - Uses exact FLBAS index without auto-detection
   - No retry on failure
   - For advanced users with specific requirements

Implementation Details:
-----------------------
- Added get_supported_lba_formats(): Queries device LBAF array via
  'nvme id-ns' JSON output, with fallback to common formats
- Added get_optimal_flbas(): Intelligent FLBAS selection prioritizing
  index 0, then formats without metadata, then common block sizes
- Enhanced create_one_ns() with retry logic and comprehensive logging
- Added proper error handling and user-friendly error messages
- Improved namespace verification with integer type conversion

Testing:
--------
- Tested on devices with FLBAS=0 (backward compatibility verified)
- Tested on devices requiring non-zero FLBAS (auto-detection works)
- Verified proper namespace creation, attachment, and verification
- Confirmed timing delays allow kernel state synchronization

Benefits:
---------
- Eliminates manual FLBAS configuration for most use cases
- Maintains full backward compatibility with existing tests
- Provides flexibility for advanced users and edge cases
- Improves test reliability across diverse NVMe hardware
- Reduces test maintenance burden for non-standard devices

### CODE REVIEW

**Comment by maramsmurthy** [comment]
[debug.log](https://github.com/user-attachments/files/27177756/debug.log)
No issues observed with changes

 (01/16) nvmetest.py:NVMeTest.test_firmware_upgrade;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (01/16) nvmetest.py:NVMeTest.test_firmware_upgrade;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: CANCEL: firmware url not given (0.34 s)
 (02/16) nvmetest.py:NVMeTest.test_create_max_ns;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (02/16) nvmetest.py:NVMeTest.test_create_max_ns;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (4.41 s)
 (03/16) nvmetest.py:NVMeTest.test_create_full_capacity_ns;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (03/16) nvmetest.py:NVMeTest.test_create_full_capacity_ns;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (12.81 s)
 (04/16) nvmetest.py:NVMeTest.testformatnamespace;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (04/16) nvmetest.py:NVMeTest.testformatnamespace;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (10.92 s)
 (05/16) nvmetest.py:NVMeTest.testread;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (05/16) nvmetest.py:NVMeTest.testread;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (0.35 s)
 (06/16) nvmetest.py:NVMeTest.testwrite;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (06/16) nvmetest.py:NVMeTest.testwrite;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (0.36 s)
 (07/16) nvmetest.py:NVMeTest.testcompare;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (07/16) nvmetest.py:NVMeTest.testcompare;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (0.38 s)
 (08/16) nvmetest.py:NVMeTest.testflush;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (08/16) nvmetest.py:NVMeTest.testflush;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (0.35 s)
 (09/16) nvmetest.py:NVMeTest.testwritezeroes;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (09/16) nvmetest.py:NVMeTest.testwritezeroes;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (0.36 s)
 (10/16) nvmetest.py:NVMeTest.testwriteuncorrectable;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (10/16) nvmetest.py:NVMeTest.testwriteuncorrectable;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (0.37 s)
 (11/16) nvmetest.py:NVMeTest.testdsm;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (11/16) nvmetest.py:NVMeTest.testdsm;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (0.36 s)
 (12/16) nvmetest.py:NVMeTest.testreset;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (12/16) nvmetest.py:NVMeTest.testreset;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (0.52 s)
 (13/16) nvmetest.py:NVMeTest.testreset_sysfs;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (13/16) nvmetest.py:NVMeTest.testreset_sysfs;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (0.53 s)
 (14/16) nvmetest.py:NVMeTest.testsubsystemreset;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (14/16) nvmetest.py:NVMeTest.testsubsystemreset;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: CANCEL: NVM Subsystem Reset is not supported (0.37 s)
 (15/16) nvmetest.py:NVMeTest.test_create_namespaces;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (15/16) nvmetest.py:NVMeTest.test_create_namespaces;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (12.96 s)
 (16/16) nvmetest.py:NVMeTest.test_delete_all_ns;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: STARTED
 (16/16) nvmetest.py:NVMeTest.test_delete_all_ns;run-Devices-device1-firmware_url-package-distro-nvme-cli-8be3: PASS (10.79 s)

File: io/disk/ssd/nvmetest.py [hunk 1]
```diff
@@ -284,16 +285,141 @@ def create_max_ns(self):
...
+        :param blocksize: Size of the namespace in blocks
+        :param controller: Controller ID to attach the namespace to
+        :param flbas: FLBAS index to use:
+                      - 0 (default): Try FLBAS=0, auto-retry on failure
+                      - -1: Skip FLBAS=0, immediately use auto-detected optimal value
+                      - 1-15: Use explicit FLBAS index, no auto-detection
+        """
+        device_name = self.device.split("/")[-1]
+        
+        # Determine retry strategy based on flbas parameter
```
Comment (sacsant) [inline io/disk/ssd/nvmetest.py]: remove this comment.

File: io/disk/ssd/nvmetest.py [hunk 2]
```diff
@@ -284,16 +285,141 @@ def create_max_ns(self):
...
+                        )
+                else:
+                    self.fail(
+                        f"Namespace creation failed with FLBAS=0 and auto-detection "
+                        f"also suggests FLBAS=0. Command: {cmd}"
+                    )
+            except Exception as e:
+                self.fail(f"Namespace creation failed: {e}")
+        elif result != 0:
+            # Failed with user-specified FLBAS (not 0), don't retry
```
Comment (sacsant) [inline io/disk/ssd/nvmetest.py]: Remove this comment, it is obvious test has failed.

File: io/disk/ssd/nvmetest.py [hunk 3]
```diff
@@ -284,16 +285,141 @@ def create_max_ns(self):
...
+                f"Namespace create command failed with FLBAS={flbas}. "
+                f"Command: {cmd}. Exit code: {result}"
+            )
+        
+        # Perform namespace rescan before attaching
+        # This is critical after namespace creation, especially after retry with different FLBAS
+        rescan_cmd = "%s ns-rescan %s" % (self.binary, self.device)
+        process.system(rescan_cmd, shell=True, ignore_status=True)
+        
+        # Small delay to allow system to recognize the new namespace
```
Comment (sacsant) [inline io/disk/ssd/nvmetest.py]: remove this comment.

File: io/disk/ssd/nvmetest.py [hunk 4]
```diff
@@ -284,16 +285,141 @@ def create_max_ns(self):
...
+        
+        # Attach namespace to controller
         cmd = "%s attach-ns %s --namespace-id=%s -controllers=%s" % (
             self.binary, self.device, ns_id, controller)
-        process.system(cmd, shell=True, ignore_status=True)
+        result = process.system(cmd, shell=True, ignore_status=True)
+        if result != 0:
+            self.fail(f"Namespace attach command failed: {cmd}. Exit code: {result}")
+        
+        # Rescan again after attachment to ensure namespace is visible
```
Comment (sacsant) [inline io/disk/ssd/nvmetest.py]: Remove the comment, it is obvious by reading the following code.

File: io/disk/ssd/nvmetest.py [hunk 5]
```diff
@@ -284,16 +285,141 @@ def create_max_ns(self):
...
+        process.system(rescan_cmd, shell=True, ignore_status=True)
+        time.sleep(2)
+        
+        # Verify namespace is now visible
```
Comment (sacsant) [inline io/disk/ssd/nvmetest.py]: Same.

File: io/disk/ssd/nvmetest.py [hunk 6]
```diff
@@ -284,16 +285,141 @@ def create_max_ns(self):
...
+        
+        # Check if the namespace ID appears in the list
+        # Format is typically: [   0]:0x1 or [   0]:0x2, etc.
+        ns_found = False
+        ns_id_int = int(ns_id)
+        ns_id_hex = f"0x{ns_id_int:x}"
+        
+        for line in ns_output.splitlines():
+            if line.startswith('['):
+                # Extract the namespace ID after the colon
```
Comment (sacsant) [inline io/disk/ssd/nvmetest.py]: This one as well.

File: io/disk/ssd/nvmetest.py [hunk 7]
```diff
@@ -21,6 +21,7 @@
 """
```
Comment (sacsant) [inline io/disk/ssd/nvmetest.py:22]: Can you shorten the commit message? Mention the problem and how this patch addresses it. Rest of the data can be added as a comment to this PR.

File: io/disk/ssd/nvmetest.py [hunk 8]
```diff
@@ -288,27 +288,27 @@ def create_max_ns(self):
     def create_one_ns(self, ns_id, blocksize, controller, flbas=0):
         """
         Creates one namespace with the specified id, block size, and controller.
-        
+
```
Comment (sacsant) [inline io/disk/ssd/nvmetest.py:292]: Can you fold this commit into previous one?

File: io/disk/ssd/nvmetest.py
Comment (sacsant) [inline io/disk/ssd/nvmetest.py:20]: Similarly, can you fold this commit into commit 2?

**Review by sacsant — CHANGES_REQUESTED** [review]

**Comment by Naresh-ibm** [comment]
@maramsmurthy as Sachin suggested, just make into two commits
blktest and nvme.  and re-push with changes requested.

**Review by Naresh-ibm — CHANGES_REQUESTED** [review]
wrap to two commits

**Comment by maramsmurthy** [comment]
@Naresh-ibm and @sacsant Can you please re-review and let me know if any other changes are required.

File: io/disk/ssd/nvmetest.py [hunk 9]
```diff
@@ -284,16 +285,121 @@ def create_max_ns(self):
...
             self.create_one_ns(str(ns_id), per_ns_blocks, ns_controller)
 
-    def create_one_ns(self, ns_id, blocksize, controller):
-        """
-        Creates one namespace, with the specified id, block size, controller
-        """
-        cmd = "%s create-ns %s --nsze=%s --ncap=%s --flbas=0 -dps=0" % (
-            self.binary, self.device, int(blocksize), int(blocksize))
-        process.system(cmd, shell=True, ignore_status=True)
+    def create_one_ns(self, ns_id, blocksize, controller, flbas=0):
```
Comment (sacsant) [inline io/disk/ssd/nvmetest.py]: create_one_ns() is only called from two places in the same file (lines 275 and 286)
Neither call passes the flbas parameter - they both use the default value of 0

The flbas=-1 code path (lines 324-330) is unreachable with current usage
The elaborate documentation about three modes is misleading because only one mode (default) is actually used

The PR adds sophisticated FLBAS detection logic, but the implementation doesn't actually leverage it properly! Will other values for FLBAS be implemented in future. If so can you add a comment. Otherwise remove the dead code.

**Review by sacsant — CHANGES_REQUESTED** [review]

File: io/disk/ssd/nvmetest.py [hunk 10]
```diff
@@ -284,16 +285,121 @@ def create_max_ns(self):
...
+
+        process.system(rescan_cmd, shell=True, ignore_status=True)
+        time.sleep(2)
+
+        ns_list_cmd = "%s list-ns %s" % (self.binary, self.device)
+        ns_output = process.system_output(ns_list_cmd, shell=True, ignore_status=True).decode("utf-8")
+
+        ns_found = False
+        ns_id_int = int(ns_id)
+        ns_id_hex = f"0x{ns_id_int:x}"
```
Comment (Naresh-ibm) [inline io/disk/ssd/nvmetest.py:426]: why do we need the hex ns_id here?

File: io/disk/ssd/nvmetest.py (hunk 9 above)
Comment (maramsmurthy) [reply io/disk/ssd/nvmetest.py]: Good catch @sacsant, As suggested will update those lines with required changes

Comment (maramsmurthy) [reply io/disk/ssd/nvmetest.py]: Remove flbas parameter input from the parameter list and changed the code to calculate the flbas parameter directly. 

New code do not assume any default values for flbas as it is not required.

File: io/disk/ssd/nvmetest.py (hunk 10 above)
Comment (maramsmurthy) [reply io/disk/ssd/nvmetest.py:426]: To match command output 

Value Conversion: The format specifier :x converts the integer 12 into its base-16 (hexadecimal) equivalent, which is c.
String Insertion: The f-string places that result directly after the literal 0x text.
Final Output: The variable ns_id_hex becomes the string 0xc, which is the command output

**Review by sacsant — CHANGES_REQUESTED** [review]
Can you merge patches 1 and 2 into one? Seems like patch 2 is incremental change to patch 1 to address review comments. 

github workflow reports pylint errors. Please review and fix it. If patch 3 is to address these changes then please merge that one also into patch 1

**Comment by maramsmurthy** [comment]
> Can you merge patches 1 and 2 into one? Seems like patch 2 is incremental change to patch 1 to address review comments.
> 
> github workflow reports pylint errors. Please review and fix it. If patch 3 is to address these changes then please merge that one also into patch 1
 
@sacsant Thank you so much for the feedback

Squashed all 3 commits into 1 and made changes.

All check are successful here. Pylint check is also successful.

Can you check and let me know if the issues persists in detail, if you are observing any

File: io/disk/ssd/nvmetest.py [hunk 11]
```diff
@@ -284,16 +286,157 @@ def create_max_ns(self):
...
+        :param ns_id: Namespace ID (typically 1-based)
+        :param blocksize: Size of namespace in blocks
+        :param controller: Controller ID to attach namespace to
         """
-        cmd = "%s create-ns %s --nsze=%s --ncap=%s --flbas=0 -dps=0" % (
-            self.binary, self.device, int(blocksize), int(blocksize))
-        process.system(cmd, shell=True, ignore_status=True)
+        flbas_value = self.get_flbas_value()
+
+        cmd = "%s create-ns %s --nsze=%s --ncap=%s --flbas=%s -dps=0" % (
```
Comment (sacsant) [inline io/disk/ssd/nvmetest.py:397]: Can you check if the name space ID should be identified after a create-ns operation?

create_max_ns() iterates over expected IDs, but those are only loop variables, not guaranteed device-assigned NSIDs. Tying attach/verification to the real created NSID avoids false failures and accidental attachment of the wrong namespace.

The new implementation calls nvme create-ns without passing a namespace identifier, then immediately attaches using the caller-supplied ns_id. On NVMe, namespace IDs are assigned by the controller and must be read back from the command output or re-queried after creation. If there are holes in the namespace table, pre-existing namespaces, or the controller chooses a different NSID than expected, attach-ns will target the wrong namespace or fail.

Rest of the code LGTM.

Comment (maramsmurthy) [reply io/disk/ssd/nvmetest.py:397]: @sacsant If we want to create max name space there should not be any other namespaces exist on the device.

So I didn't do any changes to the existing logic.

Your points are absolutely valid if we are performing multiple namespace operations on a single device.

Please let me know your opinion. If you still want to update I will go ahead with the suggested change.

Comment (sacsant) [reply io/disk/ssd/nvmetest.py:397]: > @sacsant If we want to create max name space there should not be any other namespaces exist on the device.
> 
> So I didn't do any changes to the existing logic.
> 
> Your points are absolutely valid if we are performing multiple namespace operations on a single device.
> 
> Please let me know your opinion. If you still want to update I will go ahead with the suggested change.

Thanks. Yeah, the logic change is not related to your patch. Please check this logic and if required submit a separate PR to address it.

Comment (maramsmurthy) [reply io/disk/ssd/nvmetest.py:397]: Sure Sachin! Thank you.

**Comment by maramsmurthy** [comment]
@sacsant and @Naresh-ibm 

Can you merge this if no other changes are needed for this PR

**Review by sacsant — APPROVED** [review]

**Review by Naresh-ibm — APPROVED** [review]
LGTM
